File: routes/create_container.go
// ./routes/create_container.go
package routes

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"server/method"
	"server/method/db"
	"server/structure"
	"strconv"
	"strings"

	"github.com/gofiber/fiber/v2"
)

func CreateContainer(app *fiber.App) {

	app.Post("/api/instance/create", func(c *fiber.Ctx) error {
		var data structure.CustomIhorizonData

		if err := c.BodyParser(&data); err != nil {
			fmt.Print(err)
			return c.Status(fiber.StatusBadRequest).SendString("Invalid request body")
		}

		config, err := method.LoadConfig()
		if err != nil {
			fmt.Print(err)
			return c.Status(fiber.StatusInternalServerError).SendString("Failed to load config")
		}

		if !method.ValidateDecryptedData(&data, config) {
			fmt.Println("Erreur")
			return c.Status(fiber.StatusBadRequest).SendString("Invalid data")
		}

		err = os.MkdirAll(filepath.Join(method.ProcessCWD(), "ownihrz", data.Code), os.ModePerm)
		if err != nil {
			fmt.Print(err)
			return c.Status(fiber.StatusInternalServerError).SendString("Failed to create directory")
		}

		portRange := 29268

		GitCloneCommandFormated := strings.Replace("git clone --branch {branchName} --depth 1 {repoName} .", "{branchName}", config.Container.BranchName, 1)

		cliArray := []struct {
			L   string
			CWD string
		}{
			{
				L:   strings.Replace(GitCloneCommandFormated, "{repoName}", config.Container.GithubRepo, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code),
			},

			{
				L:   "bun install",
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code),
			},

			{
				L:   "mv src/files/config.example.ts src/files/config.ts",
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code),
			},

			{
				L:   strings.Replace("sed -i 's/token: \"The bot token\"/token: \"{Auth}\"/' config.ts", "{Auth}", data.Auth, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   strings.Replace("sed -i 's/ownerid1: \"User id\",/ownerid1: \"{OwnerOne}\",/' config.ts", "{OwnerOne}", data.OwnerOne, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   strings.Replace("sed -i 's/ownerid2: \"User id\",/ownerid2: \"{OwnerTwo}\",/' config.ts", "{OwnerTwo}", data.OwnerTwo, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   strings.Replace("sed -i 's/apiToken: \"The api token\",/apiToken: \"{APIToken}\",/' config.ts", "{APIToken}", config.API.APIToken, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   "sed -i 's/useProxy: false/useProxy: true/' config.ts",
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   "sed -i 's/proxyUrl: \"https:\\/\\/login\\.domain\\.com\"/proxyUrl: \"https:\\/\\/srv\\.ihorizon\\.me\"/' config.ts",
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   strings.Replace("sed -i 's/clientSecret: \"The client secret\"/clientSecret: \"{ClientID}\"/g' config.ts", "{ClientID}", config.API.ClientID, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   strings.Replace("sed -i 's/\"3000\"/\"{PortRange}\"/' config.ts", "{PortRange}", strconv.Itoa(portRange), 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   "sed -i 's/blacklistPictureInEmbed: \"An png url\",/blacklistPictureInEmbed: \"https:\\/\\/media\\.discordapp\\.net\\/attachments\\/1099043567659384942\\/1119214828330950706\\/image\\.png\",/' config.ts",
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   strings.Replace("sed -i 's/host: \"lavalink.example.com\"/host: \"{NodeURL}\"/' config.ts", "{NodeURL}", data.Lavalink.NodeURL, '1'),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   strings.Replace("sed -i 's/authorization: \"password\"/authorization: \"{NodeAuth}\"/' config.ts", "{NodeAuth}", data.Lavalink.NodeAuth, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code, "src", "files"),
			},

			{
				L:   "npx tsc",
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code),
			},

			{
				L:   strings.Replace("mv dist/index.js dist/{Code}.js", "{Code}", data.Code, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code),
			},

			{
				L:   strings.Replace("pm2 start ./dist/{Code}.js -f", "{Code}", data.Code, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", data.Code),
			},
		}

		for _, index := range cliArray {
			cmd := exec.Command("sh", "-c", index.L)
			cmd.Dir = index.CWD
			cmd.Stdout = os.Stdout
			cmd.Stderr = os.Stderr
			if err := cmd.Run(); err != nil {
				fmt.Print(err)
				return c.Status(fiber.StatusInternalServerError).SendString(err.Error())
			}
		}

		db.Set(data.Code+"_online", true)

		if err != nil {
			return c.Status(fiber.StatusInternalServerError).SendString(err.Error())
		}

		return nil
	})

}

File: src/Routes/v1/create_container.ts
import { Custom_iHorizon } from '../../../types/OwnihrzData';
import { validateData } from '../../method/validateData.js';
import config from '../../method/getConfigData.js';
import db from '../../method/database.js';

import { Request, Response } from 'express';
import { mkdir } from "node:fs/promises";
import { execSync } from "child_process";
import path from "node:path";

export default {
    type: 'post',
    apiPath: '/api/v1/instance/create',
    run: async (req: Request, res: Response) => {

        const data = req.body as Custom_iHorizon;

        if (!config?.api.apiToken) {
            console.log("Error: Failed to load config");
            return res.status(500).send("Failed to load config");
        }

        if (!validateData(data)) {
            console.log("Error: Invalid Data");
            return res.status(403).send("Invalid Data");
        }

        await mkdir(`${process.cwd()}/ownihrz/${data.Code}`, { recursive: true });

        let port_range = 29268;

        [
            {
                l: `git clone --branch ${config.container.branchName} --depth 1 ${config.container.githubRepo} .`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code)
            },

            {
                l: `bun install`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code)
            },

            {
                l: 'cp src/files/config.example.ts src/files/config.ts',
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code)
            },

            // The bot token
            {
                l: `sed -i 's/token: "The bot token"/token: "${data.Auth}"/' config.ts`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code, 'src', 'files')
            },

            // Owner1
            {
                l: `sed -i 's/ownerid1: "User id",/ownerid1: "${data.OwnerOne}",/' config.ts`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code, 'src', 'files')
            },

            // Owner2
            {
                l: `sed -i 's/ownerid2: "User id",/ownerid2: "${data.OwnerTwo}",/' config.ts`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code, 'src', 'files'),
            },

            // ApiToken
            {
                l: `sed -i 's/apiToken: "The api token",/apiToken: "${config.api.apiToken}",/' config.ts`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code, 'src', 'files')
            },

            // Blacklist
            {
                l: `sed -i 's/blacklistPictureInEmbed: "An png url",/blacklistPictureInEmbed: "https:\\/\\/ihorizon\\.me\\/assets\\/img\\/bot\\/bsod\\.png",/' config.ts`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code, 'src', 'files')
            },

            // Lavalink Host URL
            {
                l: `sed -i 's/host: "lavalink.example.com"/host: "${data.Lavalink.NodeHost}"/' config.ts`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code, 'src', 'files')
            },

            // Lavalink Authorization
            {
                l: `sed -i 's/authorization: "password"/authorization: "${data.Lavalink.NodeAuth}"/' config.ts`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code, 'src', 'files')
            },

            // Lavalink Node Port
            {
                l: `sed -i 's/port: 2333/port: ${data.Lavalink.NodePort}/' config.ts`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code, 'src', 'files')
            },

            // Compile
            {
                l: 'bun x tsc',
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code)
            },

            // Move file
            {
                l: `mv dist/index.js dist/${data.Code}.js`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code)
            },

            // Start
            {
                l: `pm2 start ./dist/${data.Code}.js -f`,
                cwd: path.resolve(process.cwd(), 'ownihrz', data.Code)
            }
        ].forEach((index) => {
            try {
                execSync(index.l, { stdio: [0, 1, 2], cwd: index.cwd });
            } catch (e: any) {
                console.log(e.toString().split('\n')[0]);
            }
        });

        let table = db.table(`OWNIHRZ`);

        await table.set(`CLUSTER.${data.OwnerOne}.${data.Code}`,
            {
                Path: (path.resolve(process.cwd(), 'ownihrz', data.Code)) as string,
                Auth: data.Auth,
                OwnerOne: data.OwnerOne,
                OwnerTwo: data.OwnerTwo,
                Cluster: config.cluster.id,
                ExpireIn: data.ExpireIn,
                Bot: data.Bot,
                Code: data.Code
            }
        );

        res.status(200).send("CREATE !");
    },
};